ramips: add support for xiaomi RA75 Range Extender

jdeisenh opened the pull request:
This device is a close relation to the xiaomi 4A router (100m chinese
variante).
Its a 'wall wart' / Range extender form factor with one 100M LAN port
and a WPS button
Signed-off-by: Jo Deisenhofer <jo.deisenhofer@gmail.com>

jdeisenh commented:
I have a problem overwriting the LEDs definitions included from the dtsi file.
I could move the led definitions out of the dtsi files into the 3 dts that use it, but is there a better option?

jdeisenh commented:
squashed both commits into one, should clear test formalities

csharper2005 commented:
> I could move the led definitions out of the dtsi files into the 3 dts that use it, but is there a better option?

Why not?

jdeisenh commented:
I am touching other platforms that I can't test. The resulting device tree should be the same, but how can I check?
Otherwise, I have a patch for doing just that.

csharper2005 commented:
> I am touching other platforms that I can't test. The resulting device tree should be the same, but how can I check? Otherwise, I have a patch for doing just that.

You can do at least the build tests for the changed devices. And ask their owners to test the images. For example, on the openwrt in device support topics. Maybe someone can help.

The change is not so complex. It would be nice to make a separate (first) commit for the code moving from the dtsi.

csharper2005 commented:
And can you add the standard blocks such as device specification, installation & recovery instructions, MAC address table to the commit message?

jdeisenh commented:
I have verified that the resulting dtb of the potentially affected targets are unchanged (by using dtdiff). That should make the change reasonably safe. 
Otherwise done as suggested.

remittor commented on target/linux/ramips/dts/mt7628an_xiaomi_mi-ra75.dts:
```diff
@@ -0,0 +1,101 @@
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	macaddr_factory_4: macaddr@4 {
+		reg = <0x4 0x6>;
+	};
```
I think this should also be added:
```
	macaddr_factory_28: macaddr@28 {
		reg = <0x28 0x6>;
	};
	macaddr_factory_8004: macaddr@8004 {
		reg = <0x8004 0x6>;
	};
```
In my `factory` dump, I see the following MAC addresses:
```
0x4:    3C:CD:57:7E:A0:78
0x28:   64:64:4A:FD:2E:B3
0x8004: 3C:CD:57:7E:A0:79  
```

jdeisenh replied:
OK, I'll check with a fresh unit if these are actually used with the stock firmware, and for what. Adding them as suggested won't use them, right? 
BTW, How did you get your factory dump? Hardware adapter?

remittor replied:
> BTW, How did you get your factory dump? Hardware adapter?

A full dump was thrown to me by a user from the `4pda` forum. Where he got it from I don't know.

jdeisenh commented:
You are correct, these are indeed the MAC adresses the stock firmware is using. 2.4, eth, 5, in this order.
Let's see if I can find how to make use of it. 
I also have now a dump of the stock firmware

jdeisenh commented:
I now use the MAC adresses read from flash. Verified to be the same as with stock firmware.

981213 reviewed (changes requested):
Hi! It seems that all your indentations are 8 spaces. Could you change it to tabs instead?

jdeisenh commented:
I'll change the tabs and will also rebase and resolve conflicts. ETA after the weekend.

jdeisenh committed:
ramips: Move LED definition out of included dtsi into dts

Prepare for a new target with different led definitions that wants to
include this dtsi. The resulting dtb are unchanged, verified with dtdiff

Signed-off-by: Jo Deisenhofer <jo.deisenhofer@gmail.com>

jdeisenh committed:
ramips: add support for xiaomi RA75 Range Extender

This device is a 'Range extender' variant of the Xiaomi 4A router.
Its identical to the 100m non-intl/chinese version as much as it can run
the same firmware, differences being form factor, LEDs, WPS button
and one 100M port only.
The stock firmware differs significantly, being 'app managed only'.

Device specification
--------------------
SoC:      MT7628DAN MIPS_24KEc@580MHz 2.4G-n 2x2
WiFi:     MT7612EN 5G-ac 80MHz 2T2R
Flash:    16MB W25Q128BV
DRAM:     64MB built-in SoC
Switch:   built-in SoC
Ethernet: 1x10/100 Mbps
USB:      None
Antennas: 2 x external, non-detachable
LEDs:     2 programmable blue/amber
Buttons:  WPS and reset (hidden)
Housing:  Range Extender / Wall wart
Serial:   115200,8n1

MAC Addresses
-------------
All 3 MACs are read from flash and identical to stock.
Label MAC is WIFI 2G

Installation
------------
No HTML UI on this device, serial console only. The serial connector
is unpopulated but standard size and clearly marked. Flash from the
U-Boot shell at boot by choosing (2) and flashing the sysupgrade file
via tftp.

Recovery/Debricking procedures of the xiaomi 4A and variants should
work, but there currently is no official source for the stock firmware.

Signed-off-by: Jo Deisenhofer <jo.deisenhofer@gmail.com>

hauke commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -293,6 +293,9 @@ ramips_setup_macs()
 		;;
+	xiaomi,mi-router-ra75)
+		label_mac=$(mtd_get_mac_binary factory 0x4)
+		;;
```
Where is the `label_mac` being used?

jdeisenh replied:
It isn't, as far as I can tell. It documents what is the label MAC, which is not very useful. 
Removed and verified that MACs stay the same as with stock firmware

hauke commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -994,4 +1004,5 @@ define Device/zyxel_keenetic-extra-ii
   IMAGE/factory.bin := $$(sysupgrade_bin) | pad-to $$$$(BLOCKSIZE) | \
 	check-size | zyimage -d 6162 -v "ZyXEL Keenetic Extra II"
 endef
+
```
Please do not add an extra empty line

jdeisenh replied:
removed. Wherever that came from.

hauke commented on target/linux/ramips/dts/mt7628an_xiaomi_mi-router-4a-100m-intl.dts:
```diff
@@ -5,6 +5,28 @@
+
+		led_power_yellow: power_yellow {
+			label = "yellow:power";
+			gpios = <&gpio 44 GPIO_ACTIVE_LOW>;
```
You are adding the led entries here, but you haven't removed them from the `mt7628an_xiaomi_mi-router-4.dtsi` did you miss it there?

jdeisenh replied:
Since moving the led entries  was the point of the first stage patch, looks like I kind of messed this up in the process.
Coming back with a new patch after I have redone the dtdiff and verified nothing changes.

hauke replied:
You only deleted the alias in the `target/linux/ramips/dts/mt7628an_xiaomi_mi-router-4.dtsi` file, the leds node was not changed. 
The dtb is not changed because you overwrite the leds with the same definitions in the dts files.

zulhfreelancer commented:
@jdeisenh - will it work as repeater/range extender? Will the physical 100M port work as LAN port?

jdeisenh commented:
> @jdeisenh - will it work as repeater/range extender? Will the physical 100M port work as LAN port?
Yes and yes.  There is only one logical ethernet interface, and it is assigned to LAN in the default config.

jdeisenh commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -212,7 +212,7 @@ ramips_setup_macs()
 		lan_mac=$(mtd_get_mac_binary factory 0xe000)
 		wan_mac=$(macaddr_add "$lan_mac" 1)
-                ;;
+		;;
```
Also has been already adressed, and should now be visible from
